oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java

```diff
@@ -0,0 +1,286 @@
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package com.google.auth.oauth2;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.api.client.http.GenericUrl;
+import com.google.api.client.http.HttpRequest;
+import com.google.api.client.http.HttpRequestFactory;
+import com.google.api.client.http.HttpResponse;
+import com.google.api.client.json.GenericJson;
+import com.google.api.client.json.JsonParser;
+import com.google.auth.http.HttpTransportFactory;
+import com.google.common.annotations.VisibleForTesting;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * AWS credentials representing a 3PI for calling Google APIs.
+ *
+ * <p>By default, attempts to exchange the 3PI credential for a GCP access token.
+ */
+public class AwsCredentials extends ExternalAccountCredentials {
+
+  /**
+   * The AWS credential source. Stores data required to retrieve the AWS credential from the AWS
+   * metadata server.
+   */
+  static class AwsCredentialSource extends CredentialSource {
+
+    private String regionUrl;
+    private String url;
+    private String regionalCredentialVerificationUrl;
+
+    /**
+     * The source of the 3P credential. The credential source map must contain the `region_url`,
+     * `url, and `regional_cred_verification_url` entries.
+     *
+     * <p>The `region_url` is used to retrieve to targeted region.
+     *
+     * <p>The `url` is the metadata server URL which is used to retrieve the AWS credentials.
+     *
+     * <p>The `regional_cred_verification_url` is the regional GetCallerIdentity action URL, used to
+     * determine the account ID and its roles.
+     */
+    AwsCredentialSource(Map<String, Object> credentialSourceMap) {
+      super(credentialSourceMap);
+      this.regionUrl = (String) checkNotNull(credentialSourceMap.get("region_url"));
```

Comment:
NullPointerException might not be right here if the problem is an element is missing. IllegalArgumentException would be better but there might be something better than that.